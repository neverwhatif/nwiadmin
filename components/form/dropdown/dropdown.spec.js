import React from 'react';
import { shallow } from 'enzyme';

import Dropdown from './index';

jest.mock('./dropdownclear', () => 'DropdownClear');
jest.mock('./dropdowninput', () => 'DropdownInput');
jest.mock('./dropdownlist', () => 'DropdownList');

const data = [{
    id: 1,
    name: 'Lorem',
}, {
    id: 2,
    name: 'Ipsum',
}];

describe('Dropdown', () => {
    const component = shallow(<Dropdown name="lorem" data={data} />);
    const componentIsLoading = shallow(<Dropdown name="lorem" data={data} isLoading />);
    const componentWithInitialValue = shallow(<Dropdown name="lorem" data={data} value="1" />);

    it('should render', () => {
        expect(component).toMatchSnapshot();
        expect(componentIsLoading).toMatchSnapshot();
        expect(componentWithInitialValue).toMatchSnapshot();
    });
    it('should render loading', () => {
        expect(componentIsLoading.hasClass('rootDisabled')).toBe(true);
    });
    it('should render clear button', () => {
        expect(component.find('DropdownClear').length).toBe(0);
        component.setState({ value: 'lorem' });
        expect(component.find('DropdownClear').length).toBe(1);
    });
    it('should have initial value', () => {
        expect(componentWithInitialValue.state('value')).toBe('Lorem');
    });
    it('should update value', () => {
        componentWithInitialValue.setProps({ value: 1 });
        expect(componentWithInitialValue.state('value')).toBe('Lorem');
        componentWithInitialValue.setProps({ value: 2 });
        expect(componentWithInitialValue.state('value')).toBe('Ipsum');

        // Change it to the same as previously, to make sure nothing changes
        componentWithInitialValue.setProps({ value: 2 });
        expect(componentWithInitialValue.state('value')).toBe('Ipsum');
    });
    it('should open and close list state', () => {
        expect(component.state('isListOpen')).toBe(false);
        component.instance().openList();
        expect(component.state('isListOpen')).toBe(true);
        component.instance().closeList();
        expect(component.state('isListOpen')).toBe(false);
    });
    it('should have default onChange', () => {
        expect(Dropdown.defaultProps.onChange()).toBeNull();
    });
    it('should selectItem state', () => {
        // Use custom component to help keep track of it's state better
        const localComponent = shallow(<Dropdown name="lorem" data={data} />);
        localComponent.instance().closeList = jest.fn();

        expect(localComponent.state('value')).toBeNull();

        localComponent.instance().selectItem(data[0]);
        expect(localComponent.state('value')).toBe(data[0].name);
        expect(localComponent.instance().closeList).toHaveBeenCalledTimes(1);
    });
    it('should clearItem state', () => {
        // Use custom component to help keep track of it's state better
        const localComponent = shallow(<Dropdown name="lorem" data={data} />);
        localComponent.setState({ value: 'Lorem' });

        localComponent.instance().clearItem();
        expect(localComponent.state('value')).toBeNull();
    });
    it('should handle onChange when selecting', () => {
        // Use custom component to help keep track of it's state better
        const onChange = jest.fn();
        const localComponent = shallow(<Dropdown name="lorem" data={data} onChange={item => onChange(item)} />);
        localComponent.instance().selectItem(data[0]);

        expect(onChange).toHaveBeenCalledTimes(1);
        expect(onChange).toHaveBeenCalledWith({ target: { name: 'lorem', value: 1 } });
    });
    it('should handle onChange when clearing', () => {
        // Use custom component to help keep track of it's state better
        const onChange = jest.fn();
        const localComponent = shallow(<Dropdown name="lorem" data={data} onChange={item => onChange(item)} />);
        localComponent.instance().clearItem();

        expect(onChange).toHaveBeenCalledTimes(1);
        expect(onChange).toHaveBeenCalledWith({ target: { name: 'lorem', value: null } });
    });
    it('should mount and unmount', () => {
        const addEventListener = jest.fn();
        const removeEventListener = jest.fn();

        document.addEventListener = addEventListener;
        document.removeEventListener = removeEventListener;

        const unmountingComponent = shallow(<Dropdown name="lorem" data={data} />);
        unmountingComponent.unmount();

        expect(addEventListener).toHaveBeenCalledTimes(1);
        expect(removeEventListener).toHaveBeenCalledTimes(1);
    });
});
