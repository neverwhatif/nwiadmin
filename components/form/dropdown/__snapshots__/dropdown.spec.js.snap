// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown should render 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Dropdown
    data={
      Array [
        Object {
          "id": 1,
          "name": "Lorem",
        },
        Object {
          "id": 2,
          "name": "Ipsum",
        },
      ]
    }
    isDisabled={false}
    isLoading={false}
    label=""
    name="lorem"
    onChange={[Function]}
    placeholder="Select..."
    transformer={[Function]}
    value={null}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <DropdownInput
          isLoading={false}
          onClick={[Function]}
          text="Select..."
        />,
        <DropdownList
          data={
            Array [
              Object {
                "id": 1,
                "name": "Lorem",
              },
              Object {
                "id": 2,
                "name": "Ipsum",
              },
            ]
          }
          isOpen={false}
          onItemClick={[Function]}
        />,
        false,
      ],
      "className": "root",
    },
    "ref": [Function],
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "isLoading": false,
          "onClick": [Function],
          "text": "Select...",
        },
        "ref": null,
        "rendered": null,
        "type": "DropdownInput",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "data": Array [
            Object {
              "id": 1,
              "name": "Lorem",
            },
            Object {
              "id": 2,
              "name": "Ipsum",
            },
          ],
          "isOpen": false,
          "onItemClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": "DropdownList",
      },
      false,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <DropdownInput
            isLoading={false}
            onClick={[Function]}
            text="Select..."
          />,
          <DropdownList
            data={
              Array [
                Object {
                  "id": 1,
                  "name": "Lorem",
                },
                Object {
                  "id": 2,
                  "name": "Ipsum",
                },
              ]
            }
            isOpen={false}
            onItemClick={[Function]}
          />,
          false,
        ],
        "className": "root",
      },
      "ref": [Function],
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "isLoading": false,
            "onClick": [Function],
            "text": "Select...",
          },
          "ref": null,
          "rendered": null,
          "type": "DropdownInput",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "data": Array [
              Object {
                "id": 1,
                "name": "Lorem",
              },
              Object {
                "id": 2,
                "name": "Ipsum",
              },
            ],
            "isOpen": false,
            "onItemClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": "DropdownList",
        },
        false,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`Dropdown should render 2`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Dropdown
    data={
      Array [
        Object {
          "id": 1,
          "name": "Lorem",
        },
        Object {
          "id": 2,
          "name": "Ipsum",
        },
      ]
    }
    isDisabled={false}
    isLoading={true}
    label=""
    name="lorem"
    onChange={[Function]}
    placeholder="Select..."
    transformer={[Function]}
    value={null}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <DropdownInput
          isLoading={true}
          onClick={[Function]}
          text="Select..."
        />,
        <DropdownList
          data={
            Array [
              Object {
                "id": 1,
                "name": "Lorem",
              },
              Object {
                "id": 2,
                "name": "Ipsum",
              },
            ]
          }
          isOpen={false}
          onItemClick={[Function]}
        />,
        false,
      ],
      "className": "root rootDisabled",
    },
    "ref": [Function],
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "isLoading": true,
          "onClick": [Function],
          "text": "Select...",
        },
        "ref": null,
        "rendered": null,
        "type": "DropdownInput",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "data": Array [
            Object {
              "id": 1,
              "name": "Lorem",
            },
            Object {
              "id": 2,
              "name": "Ipsum",
            },
          ],
          "isOpen": false,
          "onItemClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": "DropdownList",
      },
      false,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <DropdownInput
            isLoading={true}
            onClick={[Function]}
            text="Select..."
          />,
          <DropdownList
            data={
              Array [
                Object {
                  "id": 1,
                  "name": "Lorem",
                },
                Object {
                  "id": 2,
                  "name": "Ipsum",
                },
              ]
            }
            isOpen={false}
            onItemClick={[Function]}
          />,
          false,
        ],
        "className": "root rootDisabled",
      },
      "ref": [Function],
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "isLoading": true,
            "onClick": [Function],
            "text": "Select...",
          },
          "ref": null,
          "rendered": null,
          "type": "DropdownInput",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "data": Array [
              Object {
                "id": 1,
                "name": "Lorem",
              },
              Object {
                "id": 2,
                "name": "Ipsum",
              },
            ],
            "isOpen": false,
            "onItemClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": "DropdownList",
        },
        false,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`Dropdown should render 3`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Dropdown
    data={
      Array [
        Object {
          "id": 1,
          "name": "Lorem",
        },
        Object {
          "id": 2,
          "name": "Ipsum",
        },
      ]
    }
    isDisabled={false}
    isLoading={false}
    label=""
    name="lorem"
    onChange={[Function]}
    placeholder="Select..."
    transformer={[Function]}
    value="1"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <DropdownInput
          isLoading={false}
          onClick={[Function]}
          text="Lorem"
        />,
        <DropdownList
          data={
            Array [
              Object {
                "id": 1,
                "name": "Lorem",
              },
              Object {
                "id": 2,
                "name": "Ipsum",
              },
            ]
          }
          isOpen={false}
          onItemClick={[Function]}
        />,
        <DropdownClear
          onClick={[Function]}
        />,
      ],
      "className": "root",
    },
    "ref": [Function],
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "isLoading": false,
          "onClick": [Function],
          "text": "Lorem",
        },
        "ref": null,
        "rendered": null,
        "type": "DropdownInput",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "data": Array [
            Object {
              "id": 1,
              "name": "Lorem",
            },
            Object {
              "id": 2,
              "name": "Ipsum",
            },
          ],
          "isOpen": false,
          "onItemClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": "DropdownList",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "onClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": "DropdownClear",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <DropdownInput
            isLoading={false}
            onClick={[Function]}
            text="Lorem"
          />,
          <DropdownList
            data={
              Array [
                Object {
                  "id": 1,
                  "name": "Lorem",
                },
                Object {
                  "id": 2,
                  "name": "Ipsum",
                },
              ]
            }
            isOpen={false}
            onItemClick={[Function]}
          />,
          <DropdownClear
            onClick={[Function]}
          />,
        ],
        "className": "root",
      },
      "ref": [Function],
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "isLoading": false,
            "onClick": [Function],
            "text": "Lorem",
          },
          "ref": null,
          "rendered": null,
          "type": "DropdownInput",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "data": Array [
              Object {
                "id": 1,
                "name": "Lorem",
              },
              Object {
                "id": 2,
                "name": "Ipsum",
              },
            ],
            "isOpen": false,
            "onItemClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": "DropdownList",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "onClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": "DropdownClear",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
